src/deluge/io/midi/device_specific/midi_device_lumi_keys.h

```diff
@@ -0,0 +1,194 @@
+#define MIDI_DEVICE_LUMI_KEYS_MPE_CHANNELS_PREFIX 0x10
+#define MIDI_DEVICE_LUMI_KEYS_MIDI_MODE_PREFIX 0x20
+#define MIDI_DEVICE_LUMI_KEYS_KEY_PREFIX 0x30
+#define MIDI_DEVICE_LUMI_KEYS_KEY_COUNT 12
+#define MIDI_DEVICE_LUMI_KEYS_SCALE_PREFIX 0x60
+#define MIDI_DEVICE_LUMI_KEYS_SCALE_COUNT 19
+
+#define MIDI_DEVICE_LUMI_KEYS_CONFIG_ROOT_COLOUR_PREFIX 0x30
+#define MIDI_DEVICE_LUMI_KEYS_CONFIG_GLOBAL_COLOUR_PREFIX 0x20
+
+#define MIDI_DEVICE_LUMI_SCALE_MAJOR 0b101010110101
+#define MIDI_DEVICE_LUMI_SCALE_MINOR 0b010110101101
+#define MIDI_DEVICE_LUMI_SCALE_HARMONIC_MINOR 0b100110101101
+#define MIDI_DEVICE_LUMI_SCALE_PENTATONIC_NEUTRAL 0b010010100101
+#define MIDI_DEVICE_LUMI_SCALE_PENTATONIC_MAJOR 0b001010010101
+#define MIDI_DEVICE_LUMI_SCALE_PENTATONIC_MINOR 0b010010101001
+#define MIDI_DEVICE_LUMI_SCALE_BLUES 0b010011101001
+#define MIDI_DEVICE_LUMI_SCALE_DORIAN 0b011010101101
+#define MIDI_DEVICE_LUMI_SCALE_PHRYGIAN 0b010110101011
+#define MIDI_DEVICE_LUMI_SCALE_LYDIAN 0b101011010101
+#define MIDI_DEVICE_LUMI_SCALE_MIXOLYDIAN 0b011010110101
+#define MIDI_DEVICE_LUMI_SCALE_LOCRIAN 0b010101101011
+#define MIDI_DEVICE_LUMI_SCALE_WHOLE_TONE 0b010101010101
+#define MIDI_DEVICE_LUMI_SCALE_ARABIC_A 0b101101101101
+#define MIDI_DEVICE_LUMI_SCALE_ARABIC_B 0b010101110101
+#define MIDI_DEVICE_LUMI_SCALE_JAPANESE 0b100011010001
+#define MIDI_DEVICE_LUMI_SCALE_RYUKYU 0b100010110001
+#define MIDI_DEVICE_LUMI_SCALE_8TONE_SPANISH 0b010101111011
+#define MIDI_DEVICE_LUMI_SCALE_CHROMATIC 0b111111111111
+
+class MIDIDeviceLumiKeys final : public MIDIDeviceUSBHosted {
+public:
+	static constexpr uint16_t lumiKeysVendorProductPairs[MIDI_DEVICE_LUMI_KEYS_VP_COUNT][2] = {{0x2af4, 0xe00}};
+
+	static constexpr uint8_t sysexManufacturer[3] = {0x00, 0x21, 0x10};
+
+	static constexpr uint8_t sysexMidiChannel[16][2] = {
+	    {0x20, 0x00}, // Ch. 1
+	    {0x40, 0x00}, {0x60, 0x00}, {0x00, 0x01}, {0x20, 0x01}, {0x40, 0x01}, {0x60, 0x01}, {0x00, 0x02}, {0x20, 0x02},
+	    {0x40, 0x02}, {0x60, 0x02}, {0x00, 0x03}, {0x20, 0x03}, {0x40, 0x03}, {0x60, 0x03}, {0x00, 0x04} // Ch. 16
+	};
+
+	enum class RootNote { C = 0, C_SHARP, D, D_SHARP, E, F, F_SHARP, G, G_SHARP, A, A_SHARP, B };
+
+	static constexpr uint8_t sysexRootNoteCodes[MIDI_DEVICE_LUMI_KEYS_KEY_COUNT][2] = {
+	    {0x03, 0x00}, // C
+	    {0x23, 0x00}, // C#
+	    {0x43, 0x00}, // D
+	    {0x63, 0x00}, // D#
+	    {0x03, 0x01}, // E
+	    {0x23, 0x01}, // F
+	    {0x43, 0x01}, // F#
+	    {0x63, 0x01}, // G
+	    {0x03, 0x02}, // G#
+	    {0x23, 0x02}, // A
+	    {0x43, 0x02}, // A#
+	    {0x63, 0x02}  // B
+	};
+
+	enum class MIDIMode { MPE = 0, MULTI, SINGLE };
+
+	static constexpr uint8_t sysexMidiModeCodes[3] = {
+	    0x20, // MPE
+	    0x00, // Multi
+	    0x40  // Single
+	};
+
+	enum class MPEZone { LOWER = 0, UPPER };
+
+	static constexpr uint8_t SysexMpeZoneCodes[2] = {
+	    0x05, // LOWER
+	    0x25  // UPPER
+	};
+
+	enum class Scale {
+		MAJOR = 0, // IONIAN
```

Comment:
can you reuse the scales from deluge? I think we have all of these from a recent PR, and also the format implies that it would support being sent an arbitrary scale anyway

Reply:
No, that's why there's a function to derive the scale based on the modeNotes.

Reply:
Ok - I'll have to reread that section of code since I guess I misunderstood the direction

Reply:
I did realize while working on this that the existing Deluge mode/scale code could probably benefit from being converted over to use bit array comparison and potentially save some space or gain some efficiency at some point. Using a whole uint8_t array to store and compare 1s and 0s is a bit...ehhhhh.

Reply:
Ah ok I understand it now, this makes sense to me